applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h

```diff
@@ -868,92 +665,70 @@ class MoveShallowWaterParticleUtility
-            //ModelPart::NodesContainerType local_list=aux[k];
-            //PointerVectorSet<ShallowParticle, IndexedObject> & list=aux[k];
-            BoundedMatrix<double, (TDim+1), 3 > pos;
-            BoundedMatrix<double, (TDim+1) , (TDim+1) > N;
-            unsigned int freeparticle=0; //we start with the first position in the particles array
-
-            //int local_id=1;
-            for(unsigned int ii=elem_partition[k]; ii<elem_partition[k+1]; ii++)
-            {
-                //const int & elem_id = ielem->Id();
-                ModelPart::ElementsContainerType::iterator ielem = ielembegin+ii;
+        ModelPart::ElementsContainerType::iterator i_elem_begin = mrModelPart.ElementsBegin();
+        #pragma omp parallel for firstprivate(results, pos, N, free_particle)
+        for (int ii = 0; ii < static_cast<int>(mrModelPart.NumberOfElements()); ++ii) {
+            ModelPart::ElementsContainerType::iterator ielem = i_elem_begin + ii;
+
+            if (results.size() != max_results)
                 results.resize(max_results);
-                //const int & elem_id = ielem->Id();
-                //ParticlePointerVector&  element_particle_pointers =  (ielem->GetValue(BED_PARTICLE_POINTERS));
-                //int & number_of_particles_in_elem=ielem->GetValue(NUMBER_OF_BED_PARTICLES);
-                int & number_of_particles_in_elem = mNumOfParticlesInElems[ii];
-                ParticlePointerVector&  element_particle_pointers =  mVectorOfParticlePointersVectors[ii];
-                if (number_of_particles_in_elem < (MinimumNumberOfParticles)) // && (ielem->GetGeometry())[0].Y()<0.10 )
-                {
-                    Geometry< Node<3> >& geom = ielem->GetGeometry();
-                    ComputeGaussPointPositionsForPreReseed(geom, pos, N);
 
-                    for (unsigned int j = 0; j < (pos.size1()); j++) // I am dropping the last one, the one in the middle of the element
+            int & number_of_particles_in_elem = mNumOfParticlesInElems[ii];
+            ParticlePointerVector&  element_particle_pointers =  mVectorOfParticlePointersVectors[ii];
+            if (number_of_particles_in_elem < (MinimumNumberOfParticles)) // && (ielem->GetGeometry())[0].Y()<0.10 )
+            {
+                GeometryType& geom = ielem->GetGeometry();
+                ComputeGaussPointPositionsForPreReseed(geom, pos, N);
+
+                for (unsigned int j = 0; j < (pos.size1()); j++) // I am dropping the last one, the one in the middle of the element
+                {
+                    bool keep_looking = true;
+                    while(keep_looking)
                     {
-                        bool keep_looking = true;
-                        while(keep_looking)
+                        if (mParticlesVector[free_particle].GetEraseFlag()==true)
                         {
-                            if (mParticlesVector[freeparticle].GetEraseFlag()==true)
+                            #pragma omp critical
```

Comment:
Just pointing that this `critical` is killing your performance (not sure if the same behavior can be alternatively achieved).

Reply:
this is the bottleneck of the algorithm...